src/CrawlObservers/CrawlObserver.php

```diff
@@ -8,6 +8,18 @@
 
 abstract class CrawlObserver
 {
+    /**
+     * Called when CrawlUrl is created to allow for modification
+     *
+     * @param   \Psr\Http\Message\UriInterface  $url
+     *
+     * @return \Psr\Http\Message\UriInterface
+     */
+    public function filterCrawlUrl(UriInterface $url): UriInterface
```

Comment:
Since the primary use case is modify the URL, let's rename this to `modifyCrawlUrl`.